presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1507,6 +1509,161 @@ private static void validateSphericalType(String function, OGCGeometry geometry,
+            sphericalExcess += sign * sphericalPathExcess(polygon, i);
+        }
+
+        return Math.abs(sphericalExcess * EARTH_RADIUS_M * EARTH_RADIUS_M);
+    }
+
+    private static double sphericalPathExcess(Polygon polygon, int pathIndex)
+    {
+        int start = polygon.getPathStart(pathIndex);
+        int end = polygon.getPathEnd(pathIndex);
+        if (end - start < 3) {
+            // A path with less than 3 points is not valid for calculating an area
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Can't calculate the area of a polygon with rings with less than 3 vertices. Input has %d", end - start));
+        }
+
+        // If our path is already closed, let's consider it to be shorter by 1 point
+        if (polygon.getPoint(start).equals(polygon.getPoint(end - 1))) {
+            end = end - 1;
+        }
+
+        // .getPoint allocates a new Point object on every call, unless we give it a point to store the result
+        Point tempOutPoint = new Point();
+
+        // In our path, the end point precedes the first one
+        polygon.getPoint(end - 1, tempOutPoint);
+        BearingAndExcessCalculator calculator = new BearingAndExcessCalculator(tempOutPoint);
+
+        for (int i = start; i < end; i++) {
+            polygon.getPoint(i, tempOutPoint);
+            calculator.processPoint(tempOutPoint);
+        }
+
+        return calculator.getSphericalExcess();
+    }
+
+    private static class BearingAndExcessCalculator
+    {
+        private static final double TWO_PI = 2 * Math.PI;
+        private static final double THREE_PI = 3 * Math.PI;
+
+        private double sphericalExcess;
+        private double courseDelta;
+
+        private boolean hasProcessedPoint;
+        private double firstInitialBearing;
+        private double previousBearing;
+
+        private double phiPreviousPoint;
+        private double cosPreviousPoint;
+        private double sinPreviousPoint;
+        private double tanPreviousPoint;
+        private double longitudePreviousPoint;
+
+        public BearingAndExcessCalculator(Point endPoint)
+        {
+            phiPreviousPoint = toRadians(endPoint.getY());
+            sinPreviousPoint = Math.sin(phiPreviousPoint);
+            cosPreviousPoint = Math.cos(phiPreviousPoint);
+            tanPreviousPoint = Math.tan(phiPreviousPoint / 2);
+            longitudePreviousPoint = toRadians(endPoint.getX());
+        }
+
+        private void processPoint(Point p)
+        {
+            double phiCurrentPoint = toRadians(p.getY());
+            double tanCurrentPoint = Math.tan(phiCurrentPoint / 2);
+            double longitudeCurrentPoint = toRadians(p.getX());
+
+            // Only process points which are different from the previous one,
+            // Otherwise calculating the bearing is not deterministic
+            if (longitudeCurrentPoint != longitudePreviousPoint || phiCurrentPoint != phiPreviousPoint) {
```

Comment:
This can happen only is polygon is invalid. Let's fail:

```
            if (longitude == previousLongitude && phi == previousPhi) {
                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Polygon is not valid: a loop has two identical consecutive points: %s, %s", point.getX(), point.getY()));
            }
```

Reply:
it could also be a precision problem with floating points, no ? is WKT limited in decimals ?

Reply:
ESRI docs suggest that WKTs produced by ESRI library are "limited to only return the precision of a geometry's coordinate to the eighth decimal position."

https://support.esri.com/en/technical-article/000010640